src/accumulate.jl

```diff
@@ -34,72 +92,60 @@ function partial_scan(op::Function, output::CuDeviceArray{T}, input::CuDeviceArr
         I = Rother[j]
         Ipre = Rpre[I[1]]
         Ipost = Rpost[I[2]]
-    end
 
     # load input into shared memory (apply `op` to have the correct type)
-    @inbounds temp[thread] = if i <= length(Rdim)
+    value = if i <= length(Rdim)
         op(neutral, input[Ipre, i, Ipost])
     else
         op(neutral, neutral)
     end
 
-    # build sum in place up the tree
-    offset = 1
-    d = threads>>1
-    while d > 0
+    init !== nothing && i == 1 && (value = op(init, value))
+
+    if shuffle
+        value = scan_warp(op, value, lane)
+    else
+        cache[thread] = value
+        value = scan_warp(op, value, lane, thread, cache)
         sync_threads()
```

Comment:
To test the non-shuffle version I manually turned it off and ran the same test suite, it was failing consistently and it seemed like there was a race contidition, after much trial and error I found adding this syncthreads made it work. This does not make sense to me and maybe I am missing something obvious.
When a warp is done computing its own sum it should be able to write to shared memory as no other warp has anything to do with the values accessed here.

Reply:
Don't you need this `sync_threads` because warp 1 is going to access shared memory written to by all the warps?